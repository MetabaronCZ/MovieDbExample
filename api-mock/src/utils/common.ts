import { Request } from 'express';

export type QueryParam = Request['query'][number];
export type Query = Record<string, QueryParam>;

const wait = (delay: number): Promise<void> => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const requestDelay = (cb: () => void): void => {
  const delay = 800 + Math.floor(400 * Math.random());
  wait(delay)
    .then(cb)
    .catch((error: unknown) => {
      console.error(error);
    });
};

// extract simple string value from Express query parameter result
export const parseQueryParam = (param: QueryParam): string => {
  const value = Array.isArray(param) ? param[0] : param;
  return 'string' === typeof value ? value : '';
};

type Entries<T extends Record<string, unknown>> = {
  readonly [K in keyof T]: [K, T[K]];
}[keyof T][];

// extract object entries, correctly typed
export const getObjectEntries = <T extends Record<string, unknown>>(
  obj: T,
): Entries<T> => {
  return Object.entries(obj) as Entries<T>;
};
