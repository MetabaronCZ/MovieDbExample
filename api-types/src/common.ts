export const perPages = [25, 50, 100] as const;
export const defaultPerPage = perPages[0];

export const sortDirections = ['ascending', 'descending'] as const;
export type SortDirection = (typeof sortDirections)[number];
