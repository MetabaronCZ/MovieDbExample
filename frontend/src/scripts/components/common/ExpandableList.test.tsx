import {
  fireEvent,
  getByRole,
  getByText,
  queryByRole,
  getAllByRole,
} from '@testing-library/dom';
import { act } from '@testing-library/react';
import { describe, expect, it } from '@jest/globals';

import { ExpandableList } from 'components/common/ExpandableList';
import { TestComponentWrapper, waitForComponent } from 'test-utils/component';

describe('components/common/ExpandableList', () => {
  it('should render properly', async () => {
    const { container } = await waitForComponent(
      <ExpandableList
        values={[
          { title: 'Item 1', url: '/item-1' },
          { title: 'Item 2', url: '/item-2' },
          { title: 'Item 3', url: '/item-3' },
        ]}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const visibleItem = getByText(container, 'Item 1');
    expect(visibleItem).toBeInTheDocument();
    expect(visibleItem.nodeName.toLowerCase()).toEqual('a');
    expect(visibleItem).toHaveAttribute('href', '/item-1');

    const moreButton = getByText(container, '2 další');
    expect(moreButton).toBeInTheDocument();
    expect(moreButton.nodeName.toLowerCase()).toEqual('button');
  });

  it('should render one item', async () => {
    const { container } = await waitForComponent(
      <ExpandableList values={[{ title: 'Item 1', url: '/item-1' }]} />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const visibleItem = getByText(container, 'Item 1');
    expect(visibleItem).toBeInTheDocument();

    const moreButton = queryByRole(container, 'button');
    expect(moreButton).not.toBeInTheDocument();
  });

  it('should render empty list', async () => {
    const { container } = await waitForComponent(
      <ExpandableList values={[]} />,
      { wrapper: TestComponentWrapper },
    );
    expect(container.innerHTML).toEqual('-');
  });

  it('should be expandable', async () => {
    const { container } = await waitForComponent(
      <ExpandableList
        values={[
          { title: 'Item 1', url: '/item-1' },
          { title: 'Item 2', url: '/item-2' },
          { title: 'Item 3', url: '/item-3' },
        ]}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const visibleItem = getByText(container, 'Item 1');
    expect(visibleItem).toBeInTheDocument();

    const moreButton = getByText(container, '2 další');
    expect(moreButton).toBeInTheDocument();

    // list is not visible
    let list = queryByRole(container, 'list');
    expect(list).not.toBeInTheDocument();

    // open expandable list
    act(() => {
      fireEvent.click(moreButton);
    });
    expect(container).toMatchSnapshot();

    list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const listItems = getAllByRole(container, 'listitem');
    expect(listItems.length).toEqual(2);

    const itemsContent = listItems.map((item) => item.textContent);
    expect(itemsContent).toEqual(['Item 2', 'Item 3']);

    // close expandable list
    act(() => {
      fireEvent.keyUp(document, { key: 'Escape' });
    });

    list = queryByRole(container, 'list');
    expect(list).not.toBeInTheDocument();
  });
});
