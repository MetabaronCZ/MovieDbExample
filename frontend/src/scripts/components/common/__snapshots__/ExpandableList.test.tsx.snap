// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/common/ExpandableList should be expandable 1`] = `
.c0 {
  display: flex;
  flex-direction: row;
  gap: 4px;
}

.c3 {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  color: inherit;
  text-decoration: underline;
}

.c3:hover {
  text-decoration: none;
}

.c3:focus {
  outline: 1px dashed #002d56;
  outline-color: inherit;
}

.c5 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c5:disabled {
  cursor: not-allowed;
}

.c6 {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  color: inherit;
  text-decoration: underline;
  display: inline-block;
}

.c6:hover {
  text-decoration: none;
}

.c6:focus {
  outline: 1px dashed #002d56;
  outline-color: inherit;
}

.c2 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.c1 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  min-width: 0;
}

.c7 {
  position: relative;
  flex: 1;
}

.c4 {
  white-space: nowrap;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2"
      title="Item 1"
    >
      <a
        class="c3"
        data-discover="true"
        href="/item-1"
      >
        Item 1
      </a>
    </div>
    <div
      class="c4"
    >
      (+
      <button
        class="c5 c6"
        type="button"
      >
        2 další
      </button>
      )
    </div>
    <div
      class="c7"
    />
  </div>
</div>
`;

exports[`components/common/ExpandableList should be expandable 2`] = `
.c0 {
  display: flex;
  flex-direction: row;
  gap: 4px;
}

.c3 {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  color: inherit;
  text-decoration: underline;
}

.c3:hover {
  text-decoration: none;
}

.c3:focus {
  outline: 1px dashed #002d56;
  outline-color: inherit;
}

.c5 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c5:disabled {
  cursor: not-allowed;
}

.c6 {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  color: inherit;
  text-decoration: underline;
  display: inline-block;
}

.c6:hover {
  text-decoration: none;
}

.c6:focus {
  outline: 1px dashed #002d56;
  outline-color: inherit;
}

.c2 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.c1 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  min-width: 0;
}

.c8 {
  list-style-type: none;
  position: absolute;
  top: 0;
  left: 0;
  width: max-content;
  max-width: 240px;
  padding: 8px 16px;
  background: #fff;
  box-shadow: 1px 1px 4px #dae0e3;
  z-index: 1000;
}

.c7 {
  position: relative;
  flex: 1;
}

.c4 {
  white-space: nowrap;
}

.c9 {
  margin-right: 8px;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2"
      title="Item 1"
    >
      <a
        class="c3"
        data-discover="true"
        href="/item-1"
      >
        Item 1
      </a>
    </div>
    <div
      class="c4"
    >
      (+
      <button
        class="c5 c6"
        type="button"
      >
        2 další
      </button>
      )
    </div>
    <div
      class="c7"
    >
      <ul
        class="c8"
      >
        <li
          class=""
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-right fa-sm c9"
            color="#002d56"
            data-icon="angle-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M278.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-160 160c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L210.7 256 73.4 118.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l160 160z"
              fill="currentColor"
            />
          </svg>
          <a
            class="c3"
            data-discover="true"
            href="/item-2"
          >
            Item 2
          </a>
        </li>
        <li
          class=""
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-right fa-sm c9"
            color="#002d56"
            data-icon="angle-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M278.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-160 160c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L210.7 256 73.4 118.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l160 160z"
              fill="currentColor"
            />
          </svg>
          <a
            class="c3"
            data-discover="true"
            href="/item-3"
          >
            Item 3
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`components/common/ExpandableList should render one item 1`] = `
.c1 {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  color: inherit;
  text-decoration: underline;
}

.c1:hover {
  text-decoration: none;
}

.c1:focus {
  outline: 1px dashed #002d56;
  outline-color: inherit;
}

.c0 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

<div>
  <div
    class="c0"
    title="Item 1"
  >
    <a
      class="c1"
      data-discover="true"
      href="/item-1"
    >
      Item 1
    </a>
  </div>
</div>
`;

exports[`components/common/ExpandableList should render properly 1`] = `
.c0 {
  display: flex;
  flex-direction: row;
  gap: 4px;
}

.c3 {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  color: inherit;
  text-decoration: underline;
}

.c3:hover {
  text-decoration: none;
}

.c3:focus {
  outline: 1px dashed #002d56;
  outline-color: inherit;
}

.c5 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c5:disabled {
  cursor: not-allowed;
}

.c6 {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  color: inherit;
  text-decoration: underline;
  display: inline-block;
}

.c6:hover {
  text-decoration: none;
}

.c6:focus {
  outline: 1px dashed #002d56;
  outline-color: inherit;
}

.c2 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.c1 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  min-width: 0;
}

.c7 {
  position: relative;
  flex: 1;
}

.c4 {
  white-space: nowrap;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2"
      title="Item 1"
    >
      <a
        class="c3"
        data-discover="true"
        href="/item-1"
      >
        Item 1
      </a>
    </div>
    <div
      class="c4"
    >
      (+
      <button
        class="c5 c6"
        type="button"
      >
        2 další
      </button>
      )
    </div>
    <div
      class="c7"
    />
  </div>
</div>
`;
